### PR TITLE
Improvements to implicits

### PR DESCRIPTION
**Pull request by odersky**
The PR contains an optimization and a generalization of implicits

Optimization: We now avoid recomputation of companion refs in many cases
Generalization: We allow implicit call-by-name parameters

There are also some improvements to internal diagnostics regarding implicits.
There a largish test case, which explores some schemes for generic programming over ADTs.

Review by @smarter, @OlivierBlanvillain ?

### CODE REVIEW

File: compiler/src/dotty/tools/dotc/printing/PlainPrinter.scala [hunk 1]
```diff
@@ -484,6 +485,23 @@ class PlainPrinter(_ctx: Context) extends Printer {
...
+      "NoImplicitMatches"
+    case result: DivergingImplicit =>
+      "Diverging Implicit"
+    case result: ShadowedImplicit =>
+      "Shadowed Implicit"
+    case result: FailedImplicit =>
+      "Failed Implicit"
+    case result: AmbiguousImplicits =>
+      "Ambiguous Implicit: " ~ toText(result.alt1) ~ " and " ~ toText(result.alt2)
+    case _ =>
```
Comment (odersky) [inline compiler/src/dotty/tools/dotc/printing/PlainPrinter.scala]: Looks like we can

**Comment by odersky** [comment]
Some current advantages of `enum` as a syntax:

 - less verbose syntax
 - gives us a robust way to find all children of a class. Even using sealed we still have problems with that because exploration can cause cycles.
 - Gives us an efficient way to create enums with many values.
 - Opens the potential to represent some enums as Java enums, which helps interop
 - Give us a lightweight way to hide implementation classes. E.g. We'd like `Cons`, `Nil` to be
   of type List (instead of a subtype) because this helps sometimes with type inference. 

Could this be done with Scala meta? I doubt it since the natural enum syntax is not currently legal Scala. Also there's a "meta discussion" to be had about Scala meta. Language workbenches are very appealing academically but often turn to nightmares in practice. Google for "The Lisp Curse" for details.

File: tests/run/generic/Shapes.scala [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+  case class Fst[+F](x: F) extends Sum[F, Nothing]
+  case class Snd[+S](x: S) extends Sum[Nothing, S]
+
+  case class Prod[+P1, +P2](fst: P1, snd: P2)
+
+  case class Singleton[SI](value: SI)
+
+  case class EnumValue[E](tag: Int)
+
+  trait shaped[SH1, SH2] extends unfolds[SH1, SH2]
```
Comment (smarter) [inline tests/run/generic/Shapes.scala]: What's the difference between shaped and unfolds?

File: compiler/src/dotty/tools/dotc/printing/PlainPrinter.scala (hunk 1 above)
Comment (odersky) [inline compiler/src/dotty/tools/dotc/printing/PlainPrinter.scala]: No, in fact we cannot. There are too many other cases we'd have to handle. E.g. ExplainedSearchResult. Sometimes it's better not to seal.

File: compiler/src/dotty/tools/dotc/parsing/Parsers.scala [hunk 3]
```diff
@@ -1765,12 +1765,8 @@ object Parsers {
               TypeTree()  // XX-METHOD-INFER
             } else {
               accept(COLON)
-              if (in.token == ARROW) {
-                if (owner.isTypeName && !(mods is Local))
-                  syntaxError(s"${if (mods is Mutable) "`var'" else "`val'"} parameters may not be call-by-name")
-                else if (imods.hasFlags)
-                  syntaxError("implicit parameters may not be call-by-name")
```
Comment (xeno-by) [inline compiler/src/dotty/tools/dotc/parsing/Parsers.scala]: Why was this prohibited in scalac?

**Comment by xeno-by** [comment]
@OlivierBlanvillain @odersky As defined in the examples, the syntax for enums cannot be implemented in scala.meta. In new-style macro annotations, we only accept annottees written using regular Scala syntax. 

At the moment, we don't have plans to turn scala.meta into a language workbench. The main goal of the project is to provide syntactic and semantic API for standard Scala to be used in developer tools and macros.

File: tests/run/generic/Color.scala [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+ */
+sealed trait Color extends Enum
+
+object Color extends EnumValues[Color](3) {
+
+  private def $new(tag: Int, name: String) = new Color {
+    def enumTag = tag
+    override def toString = name
+    registerEnumValue(this)
+  }
```
Comment (xeno-by) [inline tests/run/generic/Color.scala]: Can we have a named subclass of `Color` defined somewhere in the companion object and used here? I think that classnames of anonymous classes are quite ugly.

File: tests/run/generic/Enum.scala [hunk 5]
```diff
@@ -0,0 +1,18 @@
+package generic
+
+import Shapes.Singleton
+
+trait Enum {
+  def enumTag: Int
```
Comment (xeno-by) [inline tests/run/generic/Enum.scala]: What use cases do you envision for `enumTag`? Back then, @densh and I planned to use tags to speed up pattern matching. Do you think that'd make sense?

File: tests/run/generic/Enum.scala [hunk 6]
```diff
@@ -0,0 +1,18 @@
...
+}
+
+trait FiniteEnum extends Enum
```
Comment (xeno-by) [inline tests/run/generic/Enum.scala]: What's that?

File: tests/run/generic/Enum.scala [hunk 7]
```diff
@@ -0,0 +1,18 @@
...
+
+abstract class EnumValues[E <: Enum](numVals: Int) {
+  private var myValues = new Array[AnyRef](numVals)
+
+  def registerEnumValue(v: E) =
+    myValues(v.enumTag) = v
```
Comment (xeno-by) [inline tests/run/generic/Enum.scala]: Can we somehow restrict the visibility of this method?

File: tests/run/generic/List.scala [hunk 8]
```diff
@@ -0,0 +1,89 @@
...
+  abstract case class Cons[T](hd: T, tl: List0[T]) extends List0[T] {
+    def enumTag = 0
+  }
+  object Cons {
+    def apply[T](x: T, xs: List0[T]): List0[T] = new Cons(x, xs) {}
+    implicit def ConsShape[T]: Cons[T] `shaped` Prod[T, List0[T]] =
+      new (Cons[T] `shaped` Prod[T, List0[T]]) {
+        def toShape(x: Cons[T]) = Prod(x.hd, x.tl)
+        def fromShape(p: Prod[T, List0[T]]) = new Cons(p.fst, p.snd) {}
+      }
```
Comment (xeno-by) [inline tests/run/generic/List.scala]: @odersky Am I right in thinking that you'd like the `XXXShape` implicits to be derived by macros?

Comment (xeno-by) [inline tests/run/generic/List.scala]: And, in general, which parts of the `enum` proposal do you see implementable as macros, and what you'd like to hardcode in the compiler?

File: tests/run/generic/List.scala [hunk 9]
```diff
@@ -0,0 +1,89 @@
...
+        case x: Nil[T] => Snd(x)
+      }
+      def fromShape(x: Sum[Cons[T], Nil[T]]) = x match {
+        case Fst(c) => c
+        case Snd(n) => n
+      }
+    }
+}
+
+/** enum List[T] {
```
Comment (xeno-by) [inline tests/run/generic/List.scala]: Is the variance modifier omitted here on purpose?

File: tests/run/generic/List.scala (hunk 8 above)
Comment (odersky) [inline tests/run/generic/List.scala]: @xeno-by The `XXXShape` implicits can be done by the compiler. The part I need from macros are the implicits in Test.scala.

File: tests/run/generic/List.scala (hunk 9 above)
Comment (odersky) [inline tests/run/generic/List.scala]: There are actually two sources, one without variance for List0, the other with variance for List1.

**Pull request merged**